keps/sig-windows/1981-windows-privileged-container-support/README.md

```diff
@@ -452,94 +457,192 @@ change are understandable. This may include API specs (though not always
+
+```protobuf
 message WindowsContainerSecurityContext {
-	string run_as_username = 1;
-	string credential_spec = 2;
-  bool privileged = 3;
+  string run_as_username = 1;
+  string credential_spec = 2;
+  bool hostProcess = 3;
 }
 ```
-### Kubelet Implementation Details
 
-#### <b>Privileged Flag</b>
-There are no kubernetes API changes required.  Kubelet can use the privileged flag from SecurityContext api and pass that to the new privileged flag in the CRI layer.  
+*Note:* For alpha annotations on RunPodSandbox and CreateContainer CRI calls may be used until a version of containerd with Windows privileged container support is released.
 
-Add functionality to Kuberuntime_sandbox to split out the linux sandbox creation and add windows sandbox creation (WIP PR). 
+### Kubernetes API updates
 
-https://github.com/kubernetes/kubernetes/blob/a9f1d72e1de6450b890a0c0e451725468f54f515/pkg/kubelet/kuberuntime/kuberuntime_sandbox.go#L136 
+#### WindowsSecurityContextOptions.HostProcess Flag
 
-#### <b>Host Network Mode</b>
+A new boolean field named `hostProcess` will be added to [WindowsSecurityContextOptions](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#windowssecuritycontextoptions-v1-core).
 
-If using Windows privileged containers the Host Network Mode will also be enabled, as the pod will automatically get the host IP.  The PodSpec has a hostNetwork boolean field:
+On Windows, all containers in a pod must be privileged. Because of this behavior and because `WindowsSecurityContextOptions` already exists on both `PodSecurityContext` and `Container.SecurityContext` Windows containers will use this new field instead re-using the existing `privileged` field which only exists on `SecurityContext`.
+Additionally, the existing `privileged` field does not clearly describe what capabilities the container has (see https://github.com/kubernetes/kubernetes/issues/44503).
+Documentation will be added to clearly describe what capabilities these new "hostProcess" containers have.
 
-https://github.com/kubernetes/kubernetes/blob/d20fd4088476ec39c5ae2151b8fffaf0f4834418/pkg/kubelet/container/helpers.go#L227-L229
+Current behavior applies `PodSecurityContext.WindowsSecurityContextOptions` settings to all `Container.SecurityContext.WindowsSecurityContextOptions` unless those settings are already specified on the container. To address this the following API validation will be added:
 
-https://github.com/kubernetes/kubernetes/blob/a9f1d72e1de6450b890a0c0e451725468f54f515/pkg/kubelet/kuberuntime/kuberuntime_sandbox.go#L98 
+- If `PodSecurityContext.WindowsSecurityContextOptions.HostProcess = true` is set to true then no container in the pod sets `Container.SecurityContext.WindowsSecurityContextOptions.HostProcess = false`
+- If `PodSecurityContext.WindowsSecurityContextOptions.HostProcess` is not set then all containers in a pod must set `Container.SecurityContext.WindowsSecurityContextOptions.HostProcess = true`
```

Comment:
There are future plans to allow for pods to contain a mix of privileged and non-privileged containers. I originally suggested to enforce that the pod-wide `hostProcess` must be set but there was some feedback from Random-Liu that we probably shouldn't enforce that if we we'll eventually support mixed pods.
I'd be fine enforcing that for now.

Reply:
Would it be better to just enforce the pod-wide field is set until that those future ehancements come?

Reply:
talked offline, if we anticipate adding support for non-homogenous containers in the future, allowing pod-level `hostProcess=nil` and all containers having `hostProcess=true` is ok

Reply:
However, until we add non-homogenous support, the API server and kubelet should verify all containers have a consistent `hostProcess` setting. That means that if we ever relax this in API validation, back-level kubelets will refuse to run a non-homogenous pod rather than start some supposedly `hostProcess: false` containers as `hostProcess: true`